core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/RemoteEntriesIterator.scala

```diff
@@ -0,0 +1,97 @@
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.sql.execution.columnar.impl
+
+import scala.collection.JavaConverters._
+
+import com.gemstone.gemfire.internal.cache.{NonLocalRegionEntry, PartitionedRegion, RegionEntry, TXState, Token}
+import com.pivotal.gemfirexd.internal.engine.distributed.message.GetAllExecutorMessage
+import com.pivotal.gemfirexd.internal.engine.sql.execute.GemFireResultSet
+import io.snappydata.collection.LongObjectHashMap
+
+import org.apache.spark.sql.execution.columnar.impl.ColumnFormatEntry.{DELETE_MASK_COL_INDEX, STATROW_COL_INDEX}
+
+/**
+ * A [[ClusteredColumnIterator]] that fetches entries from a remote bucket.
+ */
+final class RemoteEntriesIterator(bucketId: Int, projection: Array[Int],
+    pr: PartitionedRegion, txState: TXState) extends ClusteredColumnIterator {
+
+  private val statsRows = {
+    val statsKeys = pr.getBucketKeys(bucketId).iterator().asScala.filter {
+      case k: ColumnFormatKey => k.columnIndex == STATROW_COL_INDEX
+      case _ => false
+    }
+    // get the full entries for all stats rows in one getAll
+    val msg = new GetAllExecutorMessage(pr, statsKeys.asInstanceOf[Iterator[AnyRef]].toArray,
+      null, null, null, null, null, null, txState, null, false, false)
+    statsKeys.zip(GemFireResultSet.callGetAllExecutorMessage(msg).asScala.toIterator)
+  }
+
+  /**
+   * Full projection including all of delta and meta-data columns (except base stats entry)
+   */
+  private val fullProjection = {
+    val projectionArray = new Array[Int](projection.length *
+        (ColumnDelta.USED_MAX_DEPTH + 1) - DELETE_MASK_COL_INDEX + STATROW_COL_INDEX)
+    for (i <- DELETE_MASK_COL_INDEX until STATROW_COL_INDEX) {
+      projectionArray(i - DELETE_MASK_COL_INDEX) = i
+    }
+    var i = STATROW_COL_INDEX - DELETE_MASK_COL_INDEX
+    for (p <- projection) {
+      projectionArray(i) = p
+      i += 1
+      for (d <- 0 until ColumnDelta.USED_MAX_DEPTH) {
+        projectionArray(i) = ColumnDelta.deltaColumnIndex(p - 1, d)
+        i += 1
+      }
+    }
+    projectionArray
+  }
+
+  private var currentStatsKey: ColumnFormatKey = _
+  private val currentValueMap = LongObjectHashMap.withExpectedSize[AnyRef](8)
+
+  override def hasNext: Boolean = statsRows.hasNext
+
+  override def next(): RegionEntry = {
+    currentValueMap.clear()
+    val p = statsRows.next()
+    currentStatsKey = p._1.asInstanceOf[ColumnFormatKey]
+    NonLocalRegionEntry.newEntry(currentStatsKey, p._2, pr, null)
+  }
+
+  override def getColumnValue(column: Int): AnyRef = {
+    if (currentValueMap.size() == 0) {
+      // fetch all the projected columns for current batch
+      val fetchKeys = fullProjection.map(c =>
+        new ColumnFormatKey(currentStatsKey.uuid, currentStatsKey.partitionId, c))
+      val msg = new GetAllExecutorMessage(pr, fetchKeys.asInstanceOf[Array[AnyRef]],
```

Comment:
For column table scan the txstate is all local. What is the point of passing a local txid to remote get message?

Reply:
My assumption is that this should also automatically take snapshot on remote node. So subsequent getAll calls should also read from the same snapshot. If it does not work correctly that way, then its a bug.

Reply:
I don't think it will work as per the expectation you are having at this point. @suranjan can you please see this and track it if needs attention.